### DISCUSSION TITLE
How to pass program parameters to Ghidra's debugger (WinDbg via GADP/TCP)

### DISCUSSION THREAD

**Discussion by oa2013** [NationalSecurityAgency/ghidra#3034] [Debugger]
Hi! I've built the debugger branch and I can't figure out how to pass program parameters to the debugger. 

Is that done through the interpreter window or before the debugger (WinDbg via GADP/TCP option) is connected? Below are my two screenshots:

![debugengine](https://user-images.githubusercontent.com/5304739/118284727-b934a680-b49e-11eb-8550-ba09e36ee85a.PNG)

![debuginterpreter](https://user-images.githubusercontent.com/5304739/118284738-bd60c400-b49e-11eb-9897-b22eff26c412.PNG)

**Comment by d-millar** [comment]
@oa2013 Hi, and there are probably a couple of answers to your question - hopefully one is the one you're looking for....

If you want to pass parameters to the program you're debugging, the easiest way is:
(1) connect to the agent you prefer
(2) in the Objects view, choose the "Launch (X)" button from the menu bar
(3) enter the executable name with full path and the parameters you wish to pass

If you want to pass parameters specific to a particular kind of launch:
(1) connect to the agent
(2) open up "Connections"
(3) select the connector you want
(4) hit the "Launch (X)" button and provide the parameters specific to that connector

If you want to pass parameters to the agent itself, e.g. you're connecting to a remote server
(1) open one of the GADP agents
(2) in the "DebugConnect options (.server)" entry, add the parameters, e.g. tcp:port=54321

Let us know if none of these do the trick and/or if you ideas about a better way to do what you want.  We're still experimenting a bit with this!

D

**Reply by oa2013** [reply]
Thanks so much for your prompt response! The first option is what I was looking for - there is a tiny UI bug that appears in Windows when I launched the Objects view. In the first screenshot below, you will see that the Objects view is collapsed and the executable path textbox is not visible. Once I expanded it, I was able to enter the program arguments. 

![before](https://user-images.githubusercontent.com/5304739/118293345-8e028500-b4a7-11eb-8f29-ab77da069656.PNG)
(The view is collapsed)

![after](https://user-images.githubusercontent.com/5304739/118293451-abcfea00-b4a7-11eb-8114-ad438e1994a8.PNG)
(The view is expanded)

Anyway, the debugger looks great and I hope it gets integrated into the main branch soon. My only wish-list item about the debugger would be to synchronize the decompiled code in the code window with the assembly code in the debugger window.

**Reply by d-millar** [reply]
Thanks for the feedback!  The window thing seems to happen all over the place in Ghidra.  I will ask around to see if there's a fix - with luck, that should be a happens-once and then the GUI remembers the resize.  Yell if that's not true, but I'll see if I can reproduce it.

Regarding synchronization, I assume we're talking about synchronization between the Decompiler and the Dynamic assembly view (versus the Static view).  Are you operating without anything in the Static view?  In any case, will investigate and put in a ticket!  

Thanks again for trying out the debugger!

**Reply by ryanmkurtz** [reply]
> and I hope it gets integrated into the main branch soon.

I just wanted to add the the debugger is currently in the master branch

**Reply by oa2013** [reply]
Regarding the syncing, here is what I mean - and I don't know if this is possible, of course - but if I have both the CodeBrowser and the Debugger windows open, I want to see the breakpoint from the debugger in the decompiled view:

![sync](https://user-images.githubusercontent.com/5304739/118297069-affe0680-b4ab-11eb-96d9-15b7a373d9cd.PNG)

If this is a possibility in the future, I would be happy to put in a feature request. Thanks so much again, guys and girls!

**Reply by d-millar** [reply]
Ah, got it!  Yeah, I don't think breakpoints show up in the decompiler no matter what you do.  Will add that to the ticket I just put in re Dynamic view not tracking in the Decompiler. 

You may have already figured this out, but you can get the CodeBrowser to track progress in the Debugger.  I mention it as it's not terribly obvious.  In the FrontEnd->Tools->Connect Tools..., you can pick any tool combination and connect events between them.

**Reply by oa2013** [reply]
Awesome! I've now connected the two tools and can see the decompiled code within the debugger. To add to that ticket, in addition to seeing the breakpoint, I'd also like to the decompile view to highlight the appropriate line of code as I use the step-through the code buttons on the left (like Visual Studio or Eclipse). 

But even without that, being able to see both the decompiled code and the debugger in one integrated window is huge - thanks for making that happen!

![connectedtools](https://user-images.githubusercontent.com/5304739/118301482-1df8fc80-b4b1-11eb-9b10-5661db05b583.PNG)

**Reply by d-millar** [reply]
Cool - will add.  When it does track, I think the little yellow blip in the decompiler is actually trying to follow the assembly.  It's not the entire line, I believe, because it's trying to suggest where in the instruction you are.  Nevertheless, an option here might be nice!
